proposals/0000-higher-roles.rst

```diff
@@ -0,0 +1,135 @@
+
+1. Try to solve ``forall a b. Coercible (ty a) (ty b)``. If solving succeeds, then
+   add no constraints on the role of the argument of ``ty``.
+
+2. Try to solve ``forall a b. Coercible a b => Coercible (ty a) (ty b)``. If solving
+   succeeds, then mark the argument of ``ty`` as having at least a representational
+   role.
+
+3. Otherwise, mark the argument of ``ty`` as having a nominal role.
+
+Effect and Interactions
+-----------------------
+1. We can now write::
+
+     data GList v a where
+       Mk :: (forall c d. Coercible c d => Coercible (v c) (v d)) => Metadata -> v a -> GList v a
+     type role GList representational representational
+
+2. You might be worried (I was) that the fact that datatype contexts are broken would
+   make this unsound. That is, even if we have ``G :: Type -> Type`` with ``type role G nominal``, one
+   can still talk about the type ``GList G Int``. However, we would never have a value
+   of that type, and so no threat to soundness can come of the type's existence.
+
+3. It would appear that the ``Mk`` constructor in this example would now have to hold
+   evidence for the implication constraint. However, it seems quite possible to detect this
+   shape of dictionary and eliminate it.
+
+4. Unlike previous versions of this proposal, this version does *not* change the status
+   of ``-XDatatypeContexts``.
+   
+Costs and Drawbacks
+-------------------
+1. This is potentially non-performant, requiring running the solver a lot during role inference.
+   However, it will happen only on types that have a datatype context. If we're worried about
+   performance, we could require users to write a role annotation to confirm that they want GHC
+   to do the extra work.
+
+2. It has not been proved sound. Even so, I would be comfortable adding this to GHC without a proof.
+   There are already 4 publications about roles!
+
+3. There is a remote chance that this would relax some roles in existing code, meaning that a library-writer's
+   abstraction barrier might be compromised. There is also a remote chance that I will quantum-tunnel
+   through the floor and break my leg on the floor below. Neither of these is likely.
+
+4. Newtypes will have to use datatype contexts, as newtype
+   constructors cannot have contexts.
+
+Alternatives
+------------
+* Instead of using a quantified constraint and invoking the solver, we could come up with a more
+  restrictive syntax for this idea, which expands out to a quantified constraint on every constructor.
+  This would mean we wouldn't run the solver during role inference. However, such a feature seems
+  sadly non-orthogonal when we have the features to express this idea already.
+
+* Previous versions of this proposal had the constraints either in the datatype context or
+  in the constructor contexts, but that decision was made at the level of the proposal.
+  Now, users can make their own decisions. This seems strictly better than either previous
+  proposal.
```

Comment:
This is not super clear. After a handful of reading I now assume that you mean: in previous versions of this proposal, the context used by role inference was fixed either to be from the datatype context or the constructor context; the proposal is more flexible than either, as you can use both.